netrd/distance/dmeasure.py

```diff
@@ -0,0 +1,282 @@
+            with w1 and w3, must sum to 1.0.
+        w3 (float):
+            weight of the third term in the calculation;
+            with w1 d w2, must sum to 1.0.
+        niter (int):
+            the alpha centralities are calculated using power iteration, with
+            this many iterations
+
+        Returns
+        -------
+
+        dist (float):
+            between 0 and 1, the D-measure distance between G1 and G2
+
+
+        Notes
+        -----
+        The default values for w1, w2, and w3 are from the original paper.
+
+
+        References
+        ----------
+
+        .. [1] Schieber, T. A. et al. Quantification of network structural
+               dissimilarities. Nat. Commun. 8, 13928 (2017).
+               https://www.nature.com/articles/ncomms13928
+
+        """
+
+        if sum([w1, w2, w3]) != 1:
+            raise ValueError("Weights must sum to one.")
+
+        G1 = ensure_undirected(G1)
+        G2 = ensure_undirected(G2)
+
+        first_term = 0
+        second_term = 0
+        third_term = 0
+
+        N1 = G1.number_of_nodes()
+        N2 = G2.number_of_nodes()
+
+        if w1 + w2 > 0:
+            g1_nnd, g1_pdfs = network_node_dispersion(G1)
+            g2_nnd, g2_pdfs = network_node_dispersion(G2)
+
+            first_term = np.sqrt(js_divergence(g1_pdfs, g2_pdfs))
+            second_term = np.abs(np.sqrt(g1_nnd) - np.sqrt(g2_nnd))
+
+        if w3 > 0:
+
+            def alpha_jsd(G1, G2):
+                """
+                Compute the Jensen-Shannon divergence between the
+                alpha-centrality probability distributions of two graphs.
+                """
+                p1 = alpha_centrality_prob(G1, niter=niter)
+                p2 = alpha_centrality_prob(G2, niter=niter)
+
+                m = max([len(p1), len(p2)])
+
+                P1 = np.zeros(m)
+                P2 = np.zeros(m)
+
+                P1[(m - len(p1)) : m] = p1
+                P2[(m - len(p2)) : m] = p2
+
+                return js_divergence(P1, P2)
+
+            G1c = nx.complement(G1)
+            G2c = nx.complement(G2)
+
+            first_jsd = alpha_jsd(G1, G2)
+            second_jsd = alpha_jsd(G1c, G2c)
+            third_term = 0.5 * (np.sqrt(first_jsd) + np.sqrt(second_jsd))
+
+        dist = w1 * first_term + w2 * second_term + w3 * third_term
+
+        self.results["components"] = (first_term, second_term, third_term)
+        self.results["weights"] = (w1, w2, w3)
+        self.results["dist"] = dist
+
+        return dist
+
+
+def shortest_path_matrix(G):
+    """
+    Return a matrix of pairwise shortest path lengths between nodes.
+
+    Parameters
+    ----------
+    G (nx.Graph): the graph in question
+
+    Returns
+    -------
+    pmat (np.ndarray): a matrix of shortest paths between nodes in G
+
+    """
+
+    N = G.number_of_nodes()
+    pmat = np.zeros((N, N)) + N
+
+    paths = nx.all_pairs_shortest_path_length(G)
+    for node_i, node_ij in paths:
```

Comment:
convert this to a one-line comprehension if possible?

Reply:
I don't see a way to avoid the double for loop without making a pretty nasty comprehension. You could get rid of the first for loop with a generator expression, maybe, but I think the code would turn out less readable.  Open to suggestions, though.